### ISSUE TITLE
How do I configure auth/middleware for docs?

### ISSUE THREAD

**Issue by wozniakty** [tiangolo/fastapi#420]
**Description**

I've got some middleware designed to take some of the path parameters and generate context for things like access, auth, etc.  But for calling the `/docs` and other FastAPI endpoints, this middleware is causing issues due to not having the relevant data.  I'd like to be able to turn off certain middleware for `/docs` but I'm not really sure how to configure that?  Any information would be great.

**Additional context**

Related question: is it possible to actually set middleware on a specific router rather than on the whole application?

**Comment by dmontagu** [comment]
It is possible to access the requested URL inside middleware via `request.url`, so you could use that to add custom middleware logic for specific endpoints. This would work best for one or two specific endpoints that need custom logic (e.g., the `/docs` endpoint), but probably wouldn't be great for maintainability if you want different middleware logic for different subsets of endpoints.

----

If you need to get different subsets of middleware on specific routers, you can also mount [subapplications](https://fastapi.tiangolo.com/tutorial/sub-applications-proxy/). So you could just create a different FastAPI application with the desired middleware, and mount that as a subapplication.

----

You can also add dependencies directly to a router, so if you don't need the logic to exist specifically as middleware (but just want it executed for every request hitting the specific router), you can add to the list of dependencies in the `router.include_router` call.

**Comment by wozniakty** [comment]
Yeah, the first 2 are stuff I've tried before but seemed a little hacky for the former, and a little overkill for the latter.  The third might be the best approach, just starting to dip my toes in on dependencies at this point though, so I'll have to give it a look.  Thanks.

**Comment by tiangolo** [comment]
Thanks for the help @dmontagu ! :rocket: :cake: 

Thanks @wozniakty for reporting back and closing the issue :heavy_check_mark:
